elasticsearch/client.go

```diff
@@ -0,0 +1,133 @@
+package elasticsearch
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	dphttp "github.com/ONSdigital/dp-net/http"
+	"github.com/ONSdigital/log.go/log"
+	awsauth "github.com/smartystreets/go-aws-auth"
+)
+
+// ServiceName elasticsearch
+const ServiceName = "elasticsearch"
+
+// Client is an ElasticSearch client containing an HTTP client to contact the elasticsearch API.
+type Client struct {
+	httpCli      dphttp.Clienter
+	url          string
+	serviceName  string
+	signRequests bool
+	indexes      []string
+}
+
+// NewClient returns a new initialised elasticsearch client with the default dp-net/http client
+func NewClient(url string, signRequests bool, maxRetries int, indexes ...string) *Client {
+	httpClient := dphttp.NewClient()
+	httpClient.SetMaxRetries(maxRetries)
+	return NewClientWithHTTPClient(url, signRequests, httpClient, indexes...)
+}
+
+// NewClientWithHTTPClient returns a new initialised elasticsearch client with the provided HTTP client
+func NewClientWithHTTPClient(url string, signRequests bool, httpCli dphttp.Clienter, indexes ...string) *Client {
+	return &Client{
+		httpCli:      httpCli,
+		url:          url,
+		serviceName:  ServiceName,
+		signRequests: signRequests,
+		indexes:      indexes,
+	}
+}
+
+//CreateIndex creates an index in elasticsearch
+func (cli *Client) CreateIndex(ctx context.Context, indexName string, indexSettings ...byte) (int, error) {
+
+	indexPath := cli.url + "/" + indexName
+	_, status, err := cli.CallElastic(ctx, indexPath, "PUT", indexSettings)
+	if err != nil {
+		return status, err
+	}
+	return status, nil
+}
+
+//DeleteIndex deletes an index in elasticsearch
+func (cli *Client) DeleteIndex(ctx context.Context, indexName string) (int, error) {
+
+	indexPath := cli.url + "/" + indexName
+	_, status, err := cli.CallElastic(ctx, indexPath, "DELETE", nil)
+	if err != nil {
+		return status, err
+	}
+	return status, nil
+}
+
+//AddDocument adds a JSON document to elasticsearch
+func (cli *Client) AddDocument(ctx context.Context, indexName, indexType, documentID string, document []byte) (int, error) {
+
+	documentPath := cli.url + "/" + indexName + "/" + indexType + "/" + documentID
+	_, status, err := cli.CallElastic(ctx, documentPath, "PUT", document)
+	if err != nil {
+		return status, err
+	}
+	return status, nil
+
+}
+
+// CallElastic builds a request to elasticsearch based on the method, path and payload
+func (cli *Client) CallElastic(ctx context.Context, path, method string, payload interface{}) ([]byte, int, error) {
```

Comment:
I wonder if maybe this function should be private (lower cased 'C') - do we want to expose it to users of the client?

Reply:
Yes I think it makes sense for it to be private.